cli/flox-watchdog/src/process.rs

```diff
@@ -0,0 +1,414 @@
+    Terminate,
+}
+
+pub trait Watcher {
+    /// Block while the watcher waits for a termination or cleanup event.
+    fn wait_for_termination(&mut self) -> Result<WaitResult, Error>;
+    /// Instructs the watcher to update the list of PIDs that it's watching
+    /// by reading the environment registry (for now).
+    fn update_watchlist(&mut self) -> Result<(), Error>;
+    /// Returns true if the watcher determines that it's time to perform
+    /// cleanup.
+    fn should_clean_up(&self) -> Result<bool, Error>;
+}
+
+/// The state that a process is in.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ProcStatus {
+    /// The process is running (or runnable, which includes "idle").
+    Running,
+    /// The process has exited, but has not been cleaned up by the parent.
+    Zombie,
+    /// Process is dead and will transition to a zombie or disappear.
+    /// Technically we shouldn't see this, but just in case:
+    /// https://unix.stackexchange.com/a/653370
+    AboutToBeZombie,
+    /// The process has terminated and been cleaned up
+    Dead,
+}
+
+#[derive(Debug)]
+pub struct PidWatcher {
+    pub original_pid: ActivationPid,
+    pub pids_watching: HashSet<ActivationPid>,
+    pub reg_path: PathBuf,
+    pub hash: String,
+    pub should_terminate_flag: Arc<AtomicBool>,
+    pub should_clean_up_flag: Arc<AtomicBool>,
+}
+
+impl PidWatcher {
+    /// Creates a new watcher that reads `/proc` to get process status.
+    pub fn new(
+        pid: ActivationPid,
+        reg_path: impl AsRef<Path>,
+        hash: impl AsRef<str>,
+        should_terminate_flag: Arc<AtomicBool>,
+        should_clean_up_flag: Arc<AtomicBool>,
+    ) -> Self {
+        Self {
+            original_pid: pid,
+            pids_watching: HashSet::new(),
+            reg_path: PathBuf::from(reg_path.as_ref()),
+            hash: String::from(hash.as_ref()),
+            should_terminate_flag,
+            should_clean_up_flag,
+        }
+    }
+
+    /// Reads the state of a process on macOS using `/bin/ps`, which can report
+    /// whether a process is a zombie. This is a stopgap until we someday use
+    /// `libproc`. Any failure is interpreted as an indication that the process
+    /// is no longer running.
+    #[allow(dead_code)]
+    fn read_pid_status_macos(pid: ActivationPid) -> ProcStatus {
+        let pid_raw: i32 = pid.into();
+        let stdout = match Command::new("/bin/ps")
+            .args(["-o", "state=", "-p"])
+            .arg(format!("{pid}"))
+            .output()
+        {
+            Ok(output) => output.stdout,
+            Err(err) => {
+                warn!(
+                    %err,
+                    pid = pid_raw,
+                    "failed while calling /bin/ps, treating as not running"
+                );
+                return ProcStatus::Dead;
+            },
+        };
+        if let Some(state) = stdout.first() {
+            match state {
+                b'Z' | b'?' => ProcStatus::Zombie,
```

Comment:
nonblocking suggestion: add a comment about what `?` is. I was just very confused for a minute because after activating my default environment my shell appeared to become a zombie. Turns out I was just using `unixtools.procps` instead of `/bin/ps` 😂